ZBT-1 and Yellow firmware update entities for Zigbee/Thread

puddly opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR implements the data update coordinator and firmware update entities for the ZBT-1 and Yellow integrations. This moves the responsibility for firmware updates out of the OpenThread Border Router addon and into Core, allowing for normal update entities to be used for both devices when configured for either Zigbee or Thread.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/connectzbt1.home-assistant.io/pull/143 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

puddly committed:
Initial implementation of hardware update model

puddly committed:
Fixes

puddly committed:
WIP: change the `homeassistant_sky_connect` integration type

puddly committed:
More fixes

puddly committed:
WIP

puddly committed:
Display firmware info in the device page

puddly committed:
Make progress more responsive

puddly committed:
WIP: Yellow

puddly committed:
Abstract the bootloader reset type

puddly committed:
Clean up comments

puddly committed:
Make the Yellow integration non-hardware

puddly committed:
Use the correct radio device for Yellow

puddly committed:
Avoid hardcoding strings

puddly committed:
Use `FIRMWARE_VERSION` within config flows

puddly committed:
Fix up unit tests

puddly committed:
Revert integration type changes

MartinHjelmare commented on homeassistant/components/homeassistant_hardware/coordinator.py:
```diff
@@ -0,0 +1,85 @@
+    async def _async_update_data(self) -> FirmwareManifest:
+        # Fetch the latest release metadata
+        async with self.session.get(
+            NABU_CASA_FIRMWARE_RELEASES_URL,
```
We need a Python client library, published on PyPI, for the requests and the models. We can either host it in the silabs firmware builder repo or in a new repo under the home-assistant-libs org.

MartinHjelmare commented on homeassistant/components/homeassistant_hardware/update.py:
```diff
@@ -0,0 +1,331 @@
+            )
+            firmware_type = ApplicationType.EZSP
+
+        self.entity_description = self.firmware_entity_descriptions[firmware_type]
```
The entity description shouldn't change during the entity lifetime. I'd expect us to pass the correct entity description to the entity when creating it. Then we also don't need to store the descriptions on the entity.

puddly replied:
The entity description changes when the firmware type changes: if you change the firmware type from Zigbee to Thread, for example. Is there a different way to handle this? I was avoiding hardcoding the branches and instead consolidated things into isolated entity description objects.

MartinHjelmare replied:
Remove the old entity and add a new entity.

puddly replied:
Do you have an example of an integration that replaces entities like this? I'm trying something like the following and it superficially works but I'm unable to find any other integration that does this:

```python
from homeassistant.helpers.reload import async_get_platform_without_config_entry

...

def _firmware_type_changed(
    self, old_type: ApplicationType | None, new_type: ApplicationType | None
) -> None:
    update_platform = async_get_platform_without_config_entry(
        self.hass, self._config_entry.domain, "update"
    )
    if update_platform is None:
        return

    # Remove the current entity when the firmware type changes
    ent_reg = er.async_get(self.hass)
    ent_reg.async_remove(self.entity_id)

    # And create a new one
    new_entity = type(self)(
        self._current_device,
        self._config_entry,
        self.coordinator,
        FIRMWARE_ENTITY_DESCRIPTIONS[new_type],
    )
    update_platform.add_entities([new_entity], update_before_add=True)

```

MartinHjelmare replied:
Can't we register a listener in the platform setup so we have access to all the needed names easily?

puddly replied:
Do you mean instead of `FIRMWARE_ENTITY_DESCRIPTIONS`? The entity description objects are a little more complex, as they contain logic for formatting the firmware version for display, the manifest JSON key (e.g. `skyconnect_zigbee_ncp` vs `yellow_zigbee_ncp`), the expected firmware type after installation, and a few other small things. Consolidating it all into the entity description seemed like the clearest way to do this.

puddly replied:
Let me know if <https://github.com/home-assistant/core/pull/138505/commits/970689f47d3944d2bce0cd05260d8962240fb014> is what you had in mind.

MartinHjelmare replied:
Yes.

MartinHjelmare commented on homeassistant/components/homeassistant_hardware/update.py:
```diff
@@ -0,0 +1,331 @@
+    async def _temporarily_stop_owning_software(
+        self, device: str
+    ) -> AsyncIterator[None]:
+        """Temporarily stop owning addons and integrations."""
```
I think we need a longer description what it means to stop owning the software.

puddly replied:
Integrations and addons that are currently using the device and running will be stopped by the context manager and then restarted once it exits.

MartinHjelmare replied:
Please add that to the docstring. It clarifies that it's the add-ons and integrations that are owning something, not that this method is stopping owning something.

puddly committed:
Rewrite hardware ownership context manager name, for clarity

puddly committed:
Create and delete entities instead of mutating the entity description

puddly committed:
Move manifest parsing logic into a new package

Pass the correct type to the firmware API library

puddly committed:
Move entity replacement into a `async_setup_entry` callback

MartinHjelmare commented on homeassistant/components/homeassistant_hardware/update.py:
```diff
@@ -0,0 +1,373 @@
+        )
+
+        self.async_on_remove(
+            async_dispatcher_connect(
```
Use `ConfigEntry.async_on_state_change` instead.

https://github.com/home-assistant/core/blob/a910fb879c9760da64f1db6d50787dbda03cab72/homeassistant/config_entries.py#L1182

MartinHjelmare commented on homeassistant/components/homeassistant_hardware/update.py:
```diff
@@ -0,0 +1,373 @@
+
+    # Subclasses provide the mapping between firmware types and entity descriptions
+    entity_description: FirmwareUpdateEntityDescription
+    firmware_entity_descriptions: dict[ApplicationType, FirmwareUpdateEntityDescription]
```
```suggestion
```

puddly commented on homeassistant/components/homeassistant_hardware/update.py:
```diff
@@ -0,0 +1,373 @@
+            sw_version = f"{firmware_name} {firmware_version}"
+
+        if self.device_entry is not None:
+            dr.async_get(self.hass).async_update_device(
```
The software version can change when we upgrade the firmware (or the firmware is upgraded externally and the stick is plugged in), without the firmware type changing. Should the entity be recreated in this case too or is directly updating the version fine?

MartinHjelmare replied:
It's ok to update the firmware like this and the state of the entity without removing the entity first.

puddly committed:
Change update entity category from "diagnostic" to "config"

puddly committed:
Have the client library handle firmware fetching

puddly committed:
Switch from dispatcher to `async_on_state_change`

puddly committed:
Remove unnecessary type annotation on base update entity

puddly committed:
Simplify state recomputation

puddly committed:
Remove device registry code, since the devices will not be visible

puddly committed:
Further simplify state computation

puddly committed:
Give the device-less update entity a more descriptive name

puddly committed:
Limit state changes to integer increments when sending firmware update progress

puddly committed:
Re-raise `HomeAssistantError` if there is a problem during flashing

puddly commented:
@MartinHjelmare I believe I've addressed all of your review comments.

I have removed all device registry code since we will not be exposing devices until the hardware integration changes are discussed further.

MartinHjelmare commented on homeassistant/components/homeassistant_hardware/update.py:
```diff
@@ -0,0 +1,328 @@
+        ):
+            self._latest_manifest = hardware_extra_data.firmware_manifest
+            self._maybe_recompute_state()
+            self.async_write_ha_state()
```
There's a default state write after the entity is added.
```suggestion
```

MartinHjelmare commented on homeassistant/components/homeassistant_hardware/update.py:
```diff
@@ -0,0 +1,328 @@
+        """Update the config entry after new firmware has been installed."""
+        raise NotImplementedError
+
+    def _maybe_recompute_state(self) -> None:
```
```suggestion
    def _update_attributes(self) -> None:
```

MartinHjelmare reviewed (commented):
The general design looks good. Please press "Ready for review" when ready for a review and merge.

puddly committed:
Remove unnecessary state write during entity creation

puddly committed:
Rename `_maybe_recompute_state` to `_update_attributes`

puddly committed:
Bump the flasher to 0.0.30

puddly committed:
Add some tests

puddly committed:
Ensure the update entity has a sensible name

puddly committed:
Initial ZBT-1 unit tests

puddly committed:
Replace `_update_config_entry_after_install` with a more explicit `_firmware_info_callback` override

puddly committed:
Write the firmware version to the config entry as well

puddly committed:
Test the hardware update platform independently

puddly committed:
Add unit tests to the Yellow and ZBT-1 integrations

puddly committed:
Load firmware info from the config entry when creating the update entity

puddly committed:
Test entity state restoration

puddly committed:
Test the reloading of integrations marked as "owning"

puddly committed:
Test installation failure cases

puddly committed:
Test firmware type change callback failure case

puddly committed:
Merge branch 'dev' into puddly/zbt1-update-coordinator

puddly commented:
CI failure is due to missing docs and an unrelated coverage issue, since I named a `"device"` constant:

<img width="788" alt="image" src="https://github.com/user-attachments/assets/6fbfa0ba-f7fc-4eb1-96e9-bca4bab610f3" />

puddly commented:
@MartinHjelmare Anything else I can do to progress this PR forward? I'd like to get it merged long before the beta period for testing.

Docs for the ZBT-1 and Yellow are a bit tricky here because the integrations themselves don't have documentation, they're maintained in separate repositories.

puddly commented on homeassistant/components/homeassistant_sky_connect/update.py:
```diff
@@ -145,6 +145,16 @@ def __init__(
         )
 
+        # Use the cached firmware info if it exists
+        if self._config_entry.data[FIRMWARE] is not None:
```
Restored state caches the list of available firmwares to download, we keep the current firmware information as part of the config entry data.

MartinHjelmare replied:
Is it not possible to save this information in the restore state extra data?

puddly replied:
We use the current firmware type outside of the update entity (i.e. during config/options flows) so I thought it would be better to move it out of the entity itself.

puddly replied:
Let me know if this is something that you'd like changed. I've addressed all the other review comments.

MartinHjelmare replied:
Ok.

puddly committed:
Address review comments

MartinHjelmare reviewed (approved):
Looks good!

puddly commented:
Thanks!

Once CI passes, is there a way to bypass the `docs-missing` check? Docs for these integrations aren't maintained with the rest: https://www.home-assistant.io/integrations/homeassistant_sky_connect/

MartinHjelmare commented:
Not sure. I can override the check and merge. Ping me when all the other checks are ok, if I haven't merged already then.
